internal/vault_reader.go

```diff
@@ -0,0 +1,183 @@
+type SecretReader interface {
+	ReadSecret(path string) (*vault.Secret, error)
+}
+
+type VaultSecretReader struct {
+	client *vault.Client
+}
+
+func (s VaultSecretReader) ReadSecret(path string) (*vault.Secret, error) {
+	secret, err := s.client.Logical().Read(path)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read Vault secret: %w", err)
+	}
+	return secret, nil
+}
+
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 . SecretStoreClient
+type SecretStoreClient interface {
+	ReadCredentials(path string) (CredentialsProvider, error)
+}
+
+type VaultClient struct {
+	Reader SecretReader
+}
+
+var ServiceAccountTokenReader = ReadServiceAccountToken
+var VaultClientTokenReader = ReadVaultClientToken
+var VaultAuthenticator = LoginToVault
+
+func (vc VaultClient) ReadCredentials(path string) (CredentialsProvider, error) {
+	secret, err := vc.Reader.ReadSecret(path)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read Vault secret: %w", err)
+	}
+
+	if secret == nil {
+		return nil, errors.New("returned Vault secret is nil")
+	}
+
+	if secret != nil && secret.Warnings != nil && len(secret.Warnings) > 0 {
+		return nil, fmt.Errorf("warnings were returned from Vault: %v", secret.Warnings)
+	}
+
+	if secret.Data == nil {
+		return nil, errors.New("returned Vault secret has a nil Data map")
+	}
+
+	if len(secret.Data) == 0 {
+		return nil, errors.New("returned Vault secret has an empty Data map")
+	}
+
+	if secret.Data["data"] == nil {
+		return nil, fmt.Errorf("returned Vault secret has a Data map that contains no value for key 'data'. Available keys are: %v", availableKeys(secret.Data))
+	}
+
+	data, ok := secret.Data["data"].(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("data type assertion failed for Vault secret of type: %T and value %#v read from path %s", secret.Data["data"], secret.Data["data"], path)
+	}
+
+	username, err := getValue("username", data)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get username from Vault secret: %w", err)
+	}
+
+	password, err := getValue("password", data)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get password from Vault secret: %w", err)
+	}
+
+	return ClusterCredentials{username: username, password: password}, nil
+}
+
+func getValue(key string, data map[string]interface{}) (string, error) {
+	result, ok := data[key].(string)
+	if !ok {
+		return "", fmt.Errorf("expected %s to be a string but is a %T", key, data[key])
+	}
+
+	return result, nil
+}
+
+func availableKeys(m map[string]interface{}) []string {
+	result := make([]string, len(m))
+	i := 0
+	for k := range m {
+		result[i] = k
+		i++
+	}
+	return result
+}
+
+func InitializeSecretStoreClient(vaultSpec *rabbitmqv1beta1.VaultSpec) (SecretStoreClient, error) {
+	// GCH TODO return to this...
+	// role := vaultSpec.Role
+	// if role == "" {
+	// 	return nil, errors.New("no role value set in Vault secret backend")
+	// }
+	role := "messaging-topology-operator"
```

Comment:
We discussed internally the pro's and con's of using a dedicated role for this operator versus using the role defined in the cluster definition (`rabbitmqv1beta1.VaultSpec.Role`) and the conclusion is that what you did is the best solution, i.e. a dedicated role.

Just for reference here was our analysis:

> **messaging topology operator having its own role:**

*Pros*:
- If we ever needed to prevent the messaging-topology-operator from operating on any cluster we would only need to remove its role from Vault
- It allows us to grant different permissions to the cluster-operator and messaging-topology-operator. For instance, the cluster-operator needs to request private key and certificate in addition to looking up secrets whereas the messaging-topology-operator only needs to lookup secrets. 

*Cons* :
- If we wanted to have a fine grained control on which cluster the messaging-topology-operator can operate on, we would have to modify the policy associated to the role so that it include/exclude secret paths the operator can access

>**messaging topology operator using the cluster's role defined in secreBackends.vault.role`** 

*Pros*:
- We can easily configure which clusters the messaging-topology-operator can operate when we declare the role. For instance, in the example below, we are granting the messaging-topology-operator to operate on the cluster `rabbitmq-1` by simply adding the messaging-topology-operator in the `bound_service_account_names` list.
```
vault_exec "vault write auth/kubernetes/role/rabbitmq-1-role \
 bound_service_account_names=rabbitmq-1,messaging-topology-operator \
 bound_service_account_namespaces=$RABBITMQ_NAMESPACE \
 policies=rabbitmq-policy ttl=24h"
```
*Cons*:
- If ever wanted to prevent messaging-topology-operator from operating on any cluster, we have to go over all the roles and remove its SA from the `bound_service_account_names`. Whereas if we used a dedicated role for the operator, it is just a matter of removing the role as I mentioned earlier.